### ISSUE TITLE
关于DBLoss中的DiceLoss的一个Bug

### ISSUE THREAD

**Issue by terrencew** [PaddlePaddle/PaddleOCR#11627]
在很多检测训练配置中，都采用了DistillationDBLoss，其中balance_loss为True且采用DiceLoss的情况，运行到balance loss的forward逻辑如下，可以看到事实上ohem没有用，当然与不采用ohem相比，反传的梯度肯定不同，但这其实相当于修改loss的scale，并不是ohem

def forward(self, pred, gt, mask=None):
       # pred是shrink map的prediction， gt是shrink map的0/1真值图
        """
        The BalanceLoss for Differentiable Binarization text detection
        args:
            pred (variable): predicted feature maps.
            gt (variable): ground truth feature maps.
            mask (variable): masked maps.
        return: (variable) balanced loss
        """
        positive = gt * mask
        negative = (1 - gt) * mask

        positive_count = int(positive.sum())
        negative_count = int(
            min(negative.sum(), positive_count * self.negative_ratio))
        loss = self.loss(pred, gt, mask=mask)  ## 注意这里采用Dice Loss时Loss是一个float number了，并不是一张FeatureMap

        if not self.balance_loss:
            return loss

        positive_loss = positive * loss
        negative_loss = negative * loss  # 这里实际上对每个Negative都是一样的，因为Dataloader里对shrink map的赋值是0/1的二值图
        negative_loss = paddle.reshape(negative_loss, shape=[-1])
        if negative_count > 0:
            sort_loss = negative_loss.sort(descending=True)  # 因此这里的排序是没有意义的，每个position值都是一样的。
            negative_loss = sort_loss[:negative_count]
            # negative_loss, _ = paddle.topk(negative_loss, k=negative_count_int)
            balance_loss = (positive_loss.sum() + negative_loss.sum()) / (
                positive_count + negative_count + self.eps)
        else:
            balance_loss = positive_loss.sum() / (positive_count + self.eps)
        if self.return_origin:
            return balance_loss, loss

        return balance_loss

**Comment by TingquanGao** [comment]
收到，我们排查一下。谢谢~

**Comment by tink2123** [comment]
现在的loss计算是魔改后的，自测在中文场景/文本数量多的场景也是work的【貌似ohem影响不大】，感兴趣的话可以在开源数据上做一个消融实验

**Comment by terrencew** [comment]
> 现在的loss计算是魔改后的，自测在中文场景/文本数量多的场景也是work的【貌似ohem影响不大】，感兴趣的话可以在开源数据上做一个消融实验

OK
